src/Fragcolor.Chainblocks.Claymore/Claymore.cs

```diff
@@ -0,0 +1,125 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright © 2022 Fragcolor Pte. Ltd. */
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Fragcolor.Chainblocks.Claymore
+{
+ public static class Claymore
+  {
+    private static readonly Lazy<Node> _node = new();
+
+    /// <summary>
+    /// Requests data corresponding to the specified <paramref name="hash"/>.
+    /// </summary>
+    /// <param name="hash"></param>
+    /// <returns>A variable containing the requested data.</returns>
+    public static Variable RequestData(string hash)
+    {
+      var node = _node.Value;
+      var bytes = new byte[32];
+      HexToBytes(hash, bytes);
+      ref var request = ref NativeMethods.clmrGetDataStart(bytes).AsRef();
+      PollStatePtr pollPtr = default;
+      try
+      {
+        node.Schedule(request._chain.chain);
+        do
+        {
+          Native.Core.Tick(node);
+          Thread.Sleep(25);
+        } while (!NativeMethods.clmrPoll(request.Chain(), out pollPtr));
+
+        var var = new Variable();
+        if (pollPtr.IsValid())
+        {
+          ref var poll = ref pollPtr.AsRef();
+          switch (poll._tag)
+          {
+            case PollStateTag.Failed:
+              // TODO
+              break;
+
+            case PollStateTag.Finished:
+              Native.Core.CloneVar(ref var.Value, ref request._result);
+              break;
+          }
+        }
+        node.Unschedule(request._chain.chain);
+
+        return var;
+      }
+      finally
+      {
+        if (pollPtr.IsValid()) NativeMethods.clmrPollFree(pollPtr);
+        NativeMethods.clmrGetDataFree(request.AsPointer());
+      }
+    }
+
+    public static async Task<Variable> RequestDataAsync(string hash)
```

Comment:
👏

Reply:
It has a big caveat though. It only works if it is fine to call `cmlrPoll()` from any thread following the one that initially created the request.

That's because, in .NET's `async await` pattern, the continuation after each `await` can be executed on any of the thread pool's thread. There are ways to force the same thread to resume (using some custom synchronization context) but it's hard to do right and it might not work well with the treading model of the consuming library/application.

So if we can guarantee that, it's ok. If not, then maybe it's better to only keep the blocking (synchronous) version.

Reply:
It seems to be working correctly now. That's something we can rethink about if we implement some kind of queue that can batch queries together.

Reply:
chainblocks can work from any thread too though.

Reply:
I suspect the issue if any is in the glue code/rust 🤔 
Chainblocks works with stackful coroutines so the "thread" or stack is always correct.